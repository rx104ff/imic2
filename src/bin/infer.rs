use std::env;
// The module is named `type`, so we use `r#type` to avoid conflict with the keyword.
use imic2::infer; 

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: cargo run --bin type_checker \"<JUDGMENT>\"");
        eprintln!("Example: cargo run --bin type_checker \"x:int |- if true then x else 0 : int\"");
        return;
    }

    let input = &args[1];

    // 1. Tokenize the input string.
    let tokens = infer::token::tokenize(input);
    
    // 2. The parser now consumes the entire judgment string (`env |- expr : type`)
    //    and returns a single, complete Judgment struct.
    let mut parser = infer::parser::Parser::new(tokens);
    let judgment = match parser.parse_judgment() {
        Ok(j) => j, // The parser now correctly returns the full Judgment.
        Err(e) => {
            eprintln!("Parsing Error: {}", e);
            return;
        }
    };

    // 3. Run the type checker on the parsed judgment.
    // The `judgment` object is now correctly instantiated and can be passed directly.
    match infer::infer::check_judgment(&judgment) {
        Ok(derivation) => {
            println!("{}", derivation);
        },
        Err(e) => eprintln!("Type Error: {}", e),
    }
}