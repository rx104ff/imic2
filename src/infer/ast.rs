use std::fmt;
use std::sync::atomic::{AtomicUsize, Ordering};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Var(pub String);

impl From<String> for Var {
    fn from(s: String) -> Self {
        Var(s)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Op {
    Add,
    Sub,
    Mul,
    Lt,
    Cons,
}


#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Int(i64),
    Bool(bool),
    Var(Var),
    Let(Var, Box<Expr>, Box<Expr>, bool),
    LetRec(Var, Var, Box<Expr>, Box<Expr>, bool),
    Fun(Var, Box<Expr>, bool),
    App(Box<Expr>, Box<Expr>, bool),
    If(Box<Expr>, Box<Expr>, Box<Expr>, bool),
    BinOp(Box<Expr>, Op, Box<Expr>, bool),
    Nil,
    Match(Box<Expr>, Box<Expr>, Var, Var, Box<Expr>, bool),
}

// --- Types and Type Variables for Unification ---

// A unique identifier for a type variable.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Copy)]
pub struct TypeVar(usize);

// A global atomic counter to ensure every type variable is unique.
static NEXT_TYPE_VAR_ID: AtomicUsize = AtomicUsize::new(0);

impl TypeVar {
    pub fn new() -> Self {
        TypeVar(NEXT_TYPE_VAR_ID.fetch_add(1, Ordering::SeqCst))
    }
}

// The Type enum now includes a variant for unknown types.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Int,
    Bool,
    Fun(Box<Type>, Box<Type>),
    List(Box<Type>),
    Var(TypeVar), // Represents an unknown type, e.g., 'a
}

pub type TypeEnv = Vec<(Var, Type)>;


#[derive(Debug, Clone, PartialEq)]
pub struct Judgment {
    pub env: TypeEnv,
    pub expr: Expr,
    pub ty: Type,
}

impl fmt::Display for Var {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl fmt::Display for Op {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Op::Add => write!(f, "+"),
            Op::Sub => write!(f, "-"),
            Op::Mul => write!(f, "*"),
            Op::Lt => write!(f, "<"),
            Op::Cons => write!(f, "::"),
        }
    }
}

impl fmt::Display for TypeVar {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Display type variables like 'a, 'b, 'c...
        let c = ((self.0 % 26) as u8 + b'a') as char;
        if self.0 >= 26 {
            write!(f, "'{}{}", c, self.0 / 26)
        } else {
            write!(f, "'{}", c)
        }
    }
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Type::Int => write!(f, "int"),
            Type::Bool => write!(f, "bool"),
            Type::Fun(t1, t2) => write!(f, "({} -> {})", t1, t2),
            Type::List(t) => write!(f, "{} list", t),
            Type::Var(tv) => write!(f, "{}", tv),
        }
    }
}

impl fmt::Display for Judgment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let env_str = self.env
            .iter()
            .map(|(v, t)| format!("{}: {}", v, t))
            .collect::<Vec<_>>()
            .join(", ");
        
        if env_str.is_empty() {
            write!(f, "|- {} : {}", self.expr, self.ty)
        } else {
            write!(f, "{} |- {} : {}", env_str, self.expr, self.ty)
        }
    }
}

impl fmt::Display for Expr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Expr::Int(n) => write!(f, "{}", n),
            Expr::Bool(b) => write!(f, "{}", b),
            Expr::Var(v) => write!(f, "{}", v.0),
            Expr::Nil => write!(f, "[]"),
            Expr::Let(x, e1, e2, is_paren) => {
                let s = format!("let {} = {} in {}", x, e1, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::LetRec(func, param, body, cont, is_paren) => {
                let s = format!("let rec {} = fun {} -> {} in {}", func, param, body, cont);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::If(c, t, e, is_paren) => {
                let s = format!("if {} then {} else {}", c, t, e);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::BinOp(e1, op, e2, is_paren) => {
                let s = format!("{} {} {}", e1, op, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            }
            Expr::Fun(p, b, is_paren) => {
                let s = format!("fun {} -> {}", p, b);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::App(e1, e2, is_paren) => {
                let s = format!("{} {}", e1, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::Match(e, nil_case, x, y, cons_case, is_paren) => {
                let s = format!("match {} with [] -> {} | {}::{} -> {}", e, nil_case, x, y, cons_case);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
        }
    }
}
