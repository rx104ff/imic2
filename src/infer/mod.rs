pub mod ast;
pub mod parser;
pub mod proof;
pub mod token;
pub mod infer;
pub mod unification;