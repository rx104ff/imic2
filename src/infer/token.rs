#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    // Keywords
    Let,
    In,
    If,
    Then,
    Else,
    Match,
    With,
    Fun,
    Rec,

    // Symbols
    Equals,     // =
    Bar,        // |
    Minus,      // -
    Plus,       // +
    Star,       // *
    Lt,         // <
    Arrow,      // ->
    Colon,      // :      <-- NEW for type annotations
    ColonColon, // ::
    Comma,      // ,
    Turnstile,  // |-

    // Literals and Identifiers
    Int(i64),
    Bool(bool),
    Ident(String),

    // Grouping and Lists
    LParen,
    RParen,
    Nil,
}

/// Tokenizes an input string into a sequence of Tokens for TypingML4.
pub fn tokenize(input: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    let mut chars = input.chars().peekable();

    while let Some(&ch) = chars.peek() {
        match ch {
            c if c.is_whitespace() => {
                chars.next(); // Consume whitespace
            }
            // Handle multi-character symbols first
            '|' => {
                chars.next();
                if chars.peek() == Some(&'-') {
                    chars.next();
                    tokens.push(Token::Turnstile);
                } else {
                    tokens.push(Token::Bar);
                }
            }
            '-' => {
                chars.next();
                if chars.peek() == Some(&'>') {
                    chars.next();
                    tokens.push(Token::Arrow);
                } else {
                    tokens.push(Token::Minus);
                }
            }
            ':' => {
                chars.next();
                if chars.peek() == Some(&':') {
                    chars.next();
                    tokens.push(Token::ColonColon);
                } else {
                    tokens.push(Token::Colon); // Handles single colon
                }
            }
            '[' => {
                chars.next();
                if chars.peek() == Some(&']') {
                    chars.next();
                    tokens.push(Token::Nil);
                } else {
                    panic!("'[' must be followed by ']' for an empty list literal.");
                }
            }
            // Handle keywords and identifiers
            c if c.is_alphabetic() => {
                let mut ident = String::new();
                while let Some(&c) = chars.peek() {
                    if c.is_alphanumeric() || c == '_' {
                        ident.push(c);
                        chars.next();
                    } else {
                        break;
                    }
                }
                let token = match ident.as_str() {
                    "let" => Token::Let,
                    "in" => Token::In,
                    "if" => Token::If,
                    "then" => Token::Then,
                    "else" => Token::Else,
                    "match" => Token::Match,
                    "with" => Token::With,
                    "fun" => Token::Fun,
                    "rec" => Token::Rec,
                    "true" | "True" => Token::Bool(true),
                    "false" | "False" => Token::Bool(false),
                    // Type names like "int", "bool", "list" and variables are all tokenized as Ident.
                    // The parser will be responsible for interpreting them correctly based on context.
                    _ => Token::Ident(ident),
                };
                tokens.push(token);
            }
            // Handle numbers
            c if c.is_ascii_digit() => {
                let mut number = String::new();
                while let Some(&c) = chars.peek() {
                    if c.is_ascii_digit() {
                        number.push(c);
                        chars.next();
                    } else {
                        break;
                    }
                }
                tokens.push(Token::Int(number.parse().unwrap()));
            }
            // Handle single-character symbols
            '=' => { chars.next(); tokens.push(Token::Equals); }
            '+' => { chars.next(); tokens.push(Token::Plus); }
            '*' => { chars.next(); tokens.push(Token::Star); }
            '<' => { chars.next(); tokens.push(Token::Lt); }
            '(' => { chars.next(); tokens.push(Token::LParen); }
            ')' => { chars.next(); tokens.push(Token::RParen); }
            ',' => { chars.next(); tokens.push(Token::Comma); }
            _ => {
                panic!("Unexpected character: {}", ch);
            }
        }
    }
    tokens
}
