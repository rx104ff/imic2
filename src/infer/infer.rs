use std::collections::HashMap;
use crate::infer::ast::{Expr, Op, Var, Type, TypeVar, TypeEnv, Judgment};
use crate::infer::proof::Derivation;
use crate::infer::unification::{unify, apply_sub, Substitution};

// The context for inference, holding substitutions and the type environment.
struct InferContext {
    sub: Substitution,
    env: TypeEnv,
}

impl InferContext {
    /// Creates a new type variable and wraps it in a Type enum.
    fn new_type_var(&self) -> Type {
        Type::Var(TypeVar::new())
    }
}

/// The main public entry point for the type inferrer.
/// It takes a judgment `Γ ⊢ e`, infers its type, and returns a full derivation proof.
pub fn infer_judgment(judgment: &Judgment) -> Result<Derivation, String> {
    let mut ctx = InferContext {
        sub: HashMap::new(),
        env: judgment.env.clone(),
    };
    // 1. Recursively infer the type and build the initial derivation tree.
    let mut derivation = infer_expr(&mut ctx, &judgment.expr)?;

    // 2. Apply the final substitutions to the entire derivation tree.
    apply_sub_to_deriv(&mut derivation, &ctx.sub);

    Ok(derivation)
}

/// Helper function to apply substitutions to a type environment.
fn apply_sub_to_env(env: &TypeEnv, sub: &Substitution) -> TypeEnv {
    env.iter()
        .map(|(var, ty)| (var.clone(), apply_sub(ty, sub)))
        .collect()
}

/// Applies substitutions to all types within a derivation tree, including environments.
fn apply_sub_to_deriv(deriv: &mut Derivation, sub: &Substitution) {
    // Apply substitutions to the conclusion type of the current derivation step.
    deriv.ty = apply_sub(&deriv.ty, sub);
    // CORRECTED: Also apply substitutions to the environment of the current step.
    deriv.env = apply_sub_to_env(&deriv.env, sub);

    // Recursively apply to all premises.
    for premise in &mut deriv.premises {
        apply_sub_to_deriv(premise, sub);
    }
}


/// The recursive helper that generates and solves type constraints,
/// while simultaneously building the derivation tree.
fn infer_expr(ctx: &mut InferContext, e: &Expr) -> Result<Derivation, String> {
    match e {
        Expr::Int(_) => Ok(Derivation {
            env: ctx.env.clone(), expr: e.clone(), ty: Type::Int,
            rule: "T-Int".to_string(), premises: vec![],
        }),
        Expr::Bool(_) => Ok(Derivation {
            env: ctx.env.clone(), expr: e.clone(), ty: Type::Bool,
            rule: "T-Bool".to_string(), premises: vec![],
        }),
        Expr::Nil => {
            let elem_type = ctx.new_type_var();
            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: Type::List(Box::new(elem_type)),
                rule: "T-Nil".to_string(), premises: vec![],
            })
        }
        Expr::Var(var) => {
            for (v, ty) in &ctx.env {
                if v == var {
                    return Ok(Derivation {
                        env: ctx.env.clone(), expr: e.clone(), ty: ty.clone(),
                        rule: "T-Var".to_string(), premises: vec![],
                    });
                }
            }
            Err(format!("Unbound variable: {}", var.0))
        }
        Expr::Fun(param, body, _) => {
            let param_ty = ctx.new_type_var();
            let mut new_env = ctx.env.clone();
            new_env.push((param.clone(), param_ty.clone()));
            
            let mut body_ctx = InferContext { sub: ctx.sub.clone(), env: new_env };
            let body_deriv = infer_expr(&mut body_ctx, body)?;
            
            ctx.sub = body_ctx.sub;
            let fun_ty = Type::Fun(Box::new(param_ty), Box::new(body_deriv.ty.clone()));
            
            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: fun_ty,
                rule: "T-Fun".to_string(), premises: vec![body_deriv],
            })
        }
        Expr::App(e1, e2, _) => {
            let d1 = infer_expr(ctx, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);

            let d2 = infer_expr(ctx, e2)?;
            let t2 = apply_sub(&d2.ty, &ctx.sub);

            let return_ty = ctx.new_type_var();
            let fun_ty = Type::Fun(Box::new(t2), Box::new(return_ty.clone()));
            
            let final_sub = unify(&t1, &fun_ty, &ctx.sub)?;
            ctx.sub = final_sub;
            
            let final_type = apply_sub(&return_ty, &ctx.sub);
            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: final_type,
                rule: "T-App".to_string(), premises: vec![d1, d2],
            })
        }
        Expr::Let(x, e1, e2, _) => {
            let d1 = infer_expr(ctx, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);

            let mut new_env = ctx.env.clone();
            new_env.push((x.clone(), t1));
            
            let mut body_ctx = InferContext { sub: ctx.sub.clone(), env: new_env };
            let d2 = infer_expr(&mut body_ctx, e2)?;
            
            ctx.sub = body_ctx.sub;
            let final_type = d2.ty.clone();
            
            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: final_type,
                rule: "T-Let".to_string(), premises: vec![d1, d2],
            })
        }
        Expr::BinOp(e1, op, e2, _) => {
            let d1 = infer_expr(ctx, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);

            let d2 = infer_expr(ctx, e2)?;
            let t2 = apply_sub(&d2.ty, &ctx.sub);
            
            let (expected_t1, expected_t2, result_ty, rule_name) = match op {
                Op::Add => (Type::Int, Type::Int, Type::Int, "T-Plus"),
                Op::Sub => (Type::Int, Type::Int, Type::Int, "T-Minus"),
                Op::Mul => (Type::Int, Type::Int, Type::Int, "T-Times"),
                Op::Lt => (Type::Int, Type::Int, Type::Bool, "T-Lt"),
                Op::Cons => {
                    let elem_type = ctx.new_type_var();
                    (elem_type.clone(), Type::List(Box::new(elem_type)), t2.clone(), "T-Cons")
                }
            };
            
            let sub1 = unify(&t1, &expected_t1, &ctx.sub)?;
            let sub2 = unify(&t2, &expected_t2, &sub1)?;
            ctx.sub = sub2;
            
            let final_type = apply_sub(&result_ty, &ctx.sub);
            Ok(Derivation {
                env: ctx.env.clone(),
                expr: e.clone(),
                ty: final_type,
                rule: rule_name.to_string(),
                premises: vec![d1, d2]
            })
        }
        Expr::If(cond, then_branch, else_branch, _) => {
            let d_cond = infer_expr(ctx, cond)?;
            let t_cond = apply_sub(&d_cond.ty, &ctx.sub);
            ctx.sub = unify(&t_cond, &Type::Bool, &ctx.sub)?;
            
            let d_then = infer_expr(ctx, then_branch)?;
            let t_then = apply_sub(&d_then.ty, &ctx.sub);

            let d_else = infer_expr(ctx, else_branch)?;
            let t_else = apply_sub(&d_else.ty, &ctx.sub);

            ctx.sub = unify(&t_then, &t_else, &ctx.sub)?;
            let final_type = apply_sub(&t_then, &ctx.sub);

            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: final_type,
                rule: "T-If".to_string(), premises: vec![d_cond, d_then, d_else],
            })
        }
        Expr::LetRec(f, x, e1, e2, _) => {
            let t1 = ctx.new_type_var();
            let t2 = ctx.new_type_var();
            let fun_ty = Type::Fun(Box::new(t1.clone()), Box::new(t2.clone()));
            
            let mut new_env1 = ctx.env.clone();
            new_env1.push((f.clone(), fun_ty.clone()));
            new_env1.push((x.clone(), t1));
            
            let mut body_ctx = InferContext { sub: ctx.sub.clone(), env: new_env1 };
            let d1 = infer_expr(&mut body_ctx, e1)?;
            let t_body = apply_sub(&d1.ty, &body_ctx.sub);
            
            ctx.sub = unify(&t_body, &t2, &body_ctx.sub)?;

            let mut new_env2 = ctx.env.clone();
            new_env2.push((f.clone(), fun_ty));
            
            let mut cont_ctx = InferContext { sub: ctx.sub.clone(), env: new_env2 };
            let d2 = infer_expr(&mut cont_ctx, e2)?;
            ctx.sub = cont_ctx.sub;
            
            let final_type = d2.ty.clone();
            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: final_type,
                rule: "T-LetRec".to_string(), premises: vec![d1, d2],
            })
        }
        Expr::Match(e1, e2, x, y, e3, _) => {
            let d1 = infer_expr(ctx, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);

            let elem_ty = ctx.new_type_var();
            let list_ty = Type::List(Box::new(elem_ty.clone()));
            ctx.sub = unify(&t1, &list_ty, &ctx.sub)?;

            let d2 = infer_expr(ctx, e2)?;
            let t_nil = apply_sub(&d2.ty, &ctx.sub);

            let mut new_env = ctx.env.clone();
            new_env.push((x.clone(), elem_ty.clone()));
            new_env.push((y.clone(), list_ty.clone()));
            let mut cons_ctx = InferContext { sub: ctx.sub.clone(), env: new_env };
            let d3 = infer_expr(&mut cons_ctx, e3)?;
            let t_cons = apply_sub(&d3.ty, &cons_ctx.sub);
            
            ctx.sub = unify(&t_nil, &t_cons, &cons_ctx.sub)?;
            let final_type = apply_sub(&t_nil, &ctx.sub);

            Ok(Derivation {
                env: ctx.env.clone(), expr: e.clone(), ty: final_type,
                rule: "T-Match".to_string(), premises: vec![d1, d2, d3],
            })
        }
    }
}
