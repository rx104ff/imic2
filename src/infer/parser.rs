use crate::infer::ast::{Expr, Var, Op, Type, TypeEnv, Judgment};
use crate::infer::token::Token;

/// A helper function to mark an expression as having been parsed inside parentheses.
/// This sets the `is_paren` flag to true for any compound expression type.
fn mark_expr_paren(expr: Expr) -> Expr {
    match expr {
        Expr::Let(v, e1, e2, _) => Expr::Let(v, e1, e2, true),
        Expr::LetRec(f, p, e1, e2, _) => Expr::LetRec(f, p, e1, e2, true),
        Expr::Fun(p, b, _) => Expr::Fun(p, b, true),
        Expr::App(e1, e2, _) => Expr::App(e1, e2, true),
        Expr::If(c, t, e, _) => Expr::If(c, t, e, true),
        Expr::BinOp(e1, op, e2, _) => Expr::BinOp(e1, op, e2, true),
        Expr::Match(e, n, h, t, c, _) => Expr::Match(e, n, h, t, c, true),
        // Atomic expressions like Int, Bool, Var, and Nil don't have a paren flag.
        other => other,
    }
}

/// A recursive descent parser for the TypingML4 language.
pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    /// The main entry point for the parser.
    /// It parses a judgment of the form `env |- expr : type`
    /// and returns the parsed Judgment struct.
    pub fn parse_judgment(&mut self) -> Result<Judgment, String> {
        let env = self.parse_type_env()?;
        self.expect(Token::Turnstile)?;
        let expr = self.parse_expr()?;
        self.expect(Token::Colon)?;
        let ty = self.parse_type()?;
        Ok(Judgment { env, expr, ty })
    }

    // --- Helper Methods ---
    fn peek(&self) -> Option<&Token> { self.tokens.get(self.pos) }
    fn advance(&mut self) { self.pos += 1; }
    fn expect(&mut self, expected: Token) -> Result<(), String> {
        match self.peek() {
            Some(token) if *token == expected => { self.advance(); Ok(()) }
            Some(token) => Err(format!("Expected token {:?}, found {:?}", expected, token)),
            None => Err(format!("Expected token {:?}, but found end of input.", expected)),
        }
    }
    fn next_var(&mut self) -> Result<Var, String> {
        match self.peek().cloned() {
            Some(Token::Ident(name)) => { self.advance(); Ok(Var(name)) }
            _ => Err("Expected an identifier.".to_string()),
        }
    }

    // --- Type Environment and Type Parsing ---
    fn parse_type_env(&mut self) -> Result<TypeEnv, String> {
        let mut env = TypeEnv::new();
        if self.peek() == Some(&Token::Turnstile) { return Ok(env); }
        loop {
            let var = self.next_var()?;
            self.expect(Token::Colon)?;
            let ty = self.parse_type()?;
            env.push((var, ty));
            if self.peek() == Some(&Token::Comma) { self.advance(); } 
            else { break; }
        }
        Ok(env)
    }

    fn parse_type(&mut self) -> Result<Type, String> {
        let mut ty = self.parse_single_type()?;
        if self.peek() == Some(&Token::Arrow) {
            self.advance();
            let return_ty = self.parse_type()?;
            ty = Type::Fun(Box::new(ty), Box::new(return_ty));
        }
        Ok(ty)
    }

    fn parse_single_type(&mut self) -> Result<Type, String> {
        let ty = match self.peek().cloned() {
            Some(Token::Ident(name)) => {
                self.advance();
                match name.as_str() {
                    "int" => Type::Int,
                    "bool" => Type::Bool,
                    _ => return Err(format!("Unknown type name '{}'", name)),
                }
            }
            Some(Token::LParen) => {
                self.advance();
                let inner_ty = self.parse_type()?;
                self.expect(Token::RParen)?;
                inner_ty
            }
            _ => return Err("Expected a type name or a parenthesized type.".to_string()),
        };
        if let Some(Token::Ident(s)) = self.peek() {
            if s == "list" {
                self.advance();
                return Ok(Type::List(Box::new(ty)));
            }
        }
        Ok(ty)
    }

    // --- Full Expression Parsing Logic ---

    /// The main entrypoint for parsing an expression.
    /// It correctly dispatches to low-precedence forms first.
    fn parse_expr(&mut self) -> Result<Expr, String> {
        match self.peek() {
            Some(Token::Let) => self.parse_let_expr(),
            Some(Token::If) => self.parse_if_expr(),
            Some(Token::Fun) => self.parse_fun_expr(),
            Some(Token::Match) => self.parse_match_expr(),
            _ => self.parse_lt(), // Default to operator expressions
        }
    }
    
    fn parse_let_expr(&mut self) -> Result<Expr, String> {
        self.advance(); // consume 'let'
        if self.peek() == Some(&Token::Rec) {
            self.advance(); // consume 'rec'
            let func_name = self.next_var()?;
            self.expect(Token::Equals)?;
            let body = self.parse_expr()?;
            self.expect(Token::In)?;
            let cont = self.parse_expr()?;
            if let Expr::Fun(param, fun_body, _) = body {
                 Ok(Expr::LetRec(func_name, param, fun_body, Box::new(cont), false))
            } else {
                Err("Expected a function definition after 'let rec ='".to_string())
            }
        } else {
            let var = self.next_var()?;
            self.expect(Token::Equals)?;
            let bound_expr = self.parse_expr()?;
            self.expect(Token::In)?;
            let cont = self.parse_expr()?;
            Ok(Expr::Let(var, Box::new(bound_expr), Box::new(cont), false))
        }
    }


    fn parse_if_expr(&mut self) -> Result<Expr, String> {
        self.advance(); // consume 'if'
        let cond = self.parse_expr()?;
        self.expect(Token::Then)?;
        let then_branch = self.parse_expr()?;
        self.expect(Token::Else)?;
        let else_branch = self.parse_expr()?;
        Ok(Expr::If(Box::new(cond), Box::new(then_branch), Box::new(else_branch), false))
    }

    fn parse_fun_expr(&mut self) -> Result<Expr, String> {
        self.advance(); // consume 'fun'
        let param = self.next_var()?;
        self.expect(Token::Arrow)?;
        let body = self.parse_expr()?;
        Ok(Expr::Fun(param, Box::new(body), false))
    }

    fn parse_match_expr(&mut self) -> Result<Expr, String> {
        self.advance(); // consume 'match'
        let expr_to_match = self.parse_expr()?;
        self.expect(Token::With)?;
        self.expect(Token::Nil)?;
        self.expect(Token::Arrow)?;
        let nil_case = self.parse_expr()?;
        self.expect(Token::Bar)?;
        let head_var = self.next_var()?;
        self.expect(Token::ColonColon)?;
        let tail_var = self.next_var()?;
        self.expect(Token::Arrow)?;
        let cons_case = self.parse_expr()?;
        Ok(Expr::Match(Box::new(expr_to_match), Box::new(nil_case), head_var, tail_var, Box::new(cons_case), false))
    }

    // --- Operator Precedence Parsing ---

    fn parse_lt(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_cons()?;
        while self.peek() == Some(&Token::Lt) {
            self.advance();
            let rhs = self.parse_cons()?;
            lhs = Expr::BinOp(Box::new(lhs), Op::Lt, Box::new(rhs), false);
        }
        Ok(lhs)
    }
    
    fn parse_cons(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_add_sub()?;
        if self.peek() == Some(&Token::ColonColon) {
            self.advance();
            let rhs = self.parse_cons()?;
            lhs = Expr::BinOp(Box::new(lhs), Op::Cons, Box::new(rhs), false);
        }
        Ok(lhs)
    }

    fn parse_add_sub(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_mul()?;
        while let Some(op_token) = self.peek() {
            let op = match op_token {
                Token::Plus => Op::Add, Token::Minus => Op::Sub, _ => break,
            };
            self.advance();
            let rhs = self.parse_mul()?;
            lhs = Expr::BinOp(Box::new(lhs), op, Box::new(rhs), false);
        }
        Ok(lhs)
    }

    fn parse_mul(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_app()?;
        while self.peek() == Some(&Token::Star) {
            self.advance();
            let rhs = self.parse_app()?;
            lhs = Expr::BinOp(Box::new(lhs), Op::Mul, Box::new(rhs), false);
        }
        Ok(lhs)
    }

    fn parse_app(&mut self) -> Result<Expr, String> {
        let mut lhs = self.parse_atom()?;
        while let Some(tok) = self.peek() {
            match tok {
                Token::Int(_) | Token::Bool(_) | Token::Ident(_) | Token::LParen | Token::Nil => {
                    let rhs = self.parse_atom()?;
                    lhs = Expr::App(Box::new(lhs), Box::new(rhs), false);
                },
                _ => break,
            }
        }
        Ok(lhs)
    }

    fn parse_atom(&mut self) -> Result<Expr, String> {
        match self.peek().cloned() {
            Some(Token::Int(n)) => { self.advance(); Ok(Expr::Int(n)) }
            Some(Token::Bool(b)) => { self.advance(); Ok(Expr::Bool(b)) }
            Some(Token::Ident(name)) => { self.advance(); Ok(Expr::Var(Var(name))) }
            Some(Token::Nil) => { self.advance(); Ok(Expr::Nil) }
            Some(Token::LParen) => {
                self.advance();
                let expr = self.parse_expr()?;
                self.expect(Token::RParen)?;
                Ok(mark_expr_paren(expr))
            }
            _ => Err("Unexpected token found at atomic level.".to_string())
        }
    }
}
