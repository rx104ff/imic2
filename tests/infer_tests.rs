use imic2::infer;
use insta::assert_snapshot;

/// A helper function to simulate a full run of the type checker.
fn run_type_test(input: &str) -> String {
    let tokens = infer::token::tokenize(input);
    let mut parser = infer::parser::Parser::new(tokens);
    let judgment = match parser.parse_judgment() {
        Ok(j) => j,
        Err(e) => return format!("Parsing Error: {}", e),
    };

    match infer::infer::infer_judgment(&judgment) {
        Ok(derivation) => format!("{}", derivation),
        Err(e) => format!("Type Error: {}", e),
    }
}

#[test]
fn test_infer_typing_80() {
    let result = run_type_test("|- 3 + 5 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_typing_81() {
    let result = run_type_test("|- if 4 < 5 then 2 + 3 else 8 * 8 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_typing_82() {
    let result = run_type_test("x : bool, y : int |- if x then y + 1 else y - 1 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_typing_83() {
    let result = run_type_test("|- let x = 3 < 2 in let y = 5 in if x then y else 2 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_typing_84() {
    let result = run_type_test("|- fun x -> x + 1 : int -> int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_typing_85() {
    let result = run_type_test("|- let f = fun x -> x + 1 in f 4 : int");
    assert_snapshot!(&result);
}
