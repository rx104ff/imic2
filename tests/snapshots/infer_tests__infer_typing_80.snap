---
source: tests/infer_tests.rs
expression: "&result"
---
|- 3 + 5 : int by T-Plus {
    |- 3 : int by T-Int {
    };
    |- 5 : int by T-Int {
    };
};
